src/init.cpp

```diff
@@ -286,6 +291,30 @@ void HandleSIGHUP(int)
     fReopenDebugLog = true;
 }
 
+#ifdef WIN32
+BOOL CtrlHandler( DWORD fdwCtrlType ) 
+{ 
+  switch( fdwCtrlType ) 
+  { 
+    case CTRL_C_EVENT: 
+    case CTRL_CLOSE_EVENT: 
+    case CTRL_BREAK_EVENT: 
+    case CTRL_LOGOFF_EVENT: 
+    case CTRL_SHUTDOWN_EVENT: 
+      fRequestShutdown = true;
+      std::cout << _("Stopping node.  This may take a while, be patient!") << std::endl;
+      while (!fShutdownCompleted)
+      {
+          MilliSleep(100);
+      }
```

Comment:
I would recommend a hard-cap on the total time spent asleep (maybe 3 seconds for a best effort?) to prevent a hanging process.